Add `max_nfev` as uniform way to specify maximum number of function evaluations

newville opened the pull request:
This is an effort (perhaps incomplete) to address #332 so that there is a uniform way to specify "maximum number of function evaluations" or "iterations" to a fit.  Currently, each `scipy.optimize` function has its own keyword expressing this constant.  

-   `scipy.optimize.minimize()` uses `maxiter`.
-   `scipy.optimize.leastsq()` uses `maxfev`.
-   `scipy.optimize.least_squares()` uses `max_nfev`.

Other functions use 'maxfun' .  Some appear to suggest that `maxfev` is an alias for `maxiter`. 
I think that 'maxiter` is confusing as a "fit iteration" is poorly defined (is that one function calls, Nvariable function calls, or something else). 

With the PR, we try to standardize on `max_nfev` to mean "maximum number of evaluations of the fit function".  To be clear, we are not guaranteeing this limit, especially **in total** -- the actual number of calls may be slightly (say up to Nvariables) more.

This uses `max_nfev` partly based on the precedence of `least_squares`, and partly because it seems pretty clear.

There is an added test, and updated docstrings so that this gets rendered to docs. 

I would like to tag 0.9.15rc1 very soon (like, within a couple days).  I am OK with trying to include this in 0.9.15, or this could wait until preparing version 1.0.   

  
<!--- What type of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix
- [x] New feature
- [ ] Refactoring / maintenance
- [ ] Documentation / examples


###### Tested on
<!-- Generate version information with this command in the Python shell and copy the output here:
import sys, lmfit, numpy, scipy, asteval, uncertainties, six
print('Python: {}\n\nlmfit: {}, scipy: {}, numpy: {}, asteval: {}, uncertainties: {}, six: {}'\
      .format(sys.version, lmfit.__version__, scipy.__version__, numpy.__version__, \
      asteval.__version__, uncertainties.__version__, six.__version__))
-->

Python: 3.7.4 (default, Aug 13 2019, 20:35:49) 
[GCC 7.3.0]

lmfit: 0.9.14+60.gd83dbc6, scipy: 1.3.1, numpy: 1.17.2, asteval: 0.9.16, uncertainties: 3.0.2, six: 1.12.0

###### Verification <!-- (delete not applicable items) -->
Have you
<!--- Put an `x` in all the boxes that apply OR describe why you think this is unnecessary. -->
- [x] included docstrings that follow PEP 257?
<!-- Please use your favorite linter (e.g., pydocstyle) to check your docstrings. -->
- [x] referenced existing Issue and/or provided relevant link to mailing list?
<!-- Please don't open a new Issue if you are submitting a pull request. -->
- [x] verified that existing tests pass locally?
<!-- Please run the test-suite locally with pytest and make sure it passes. -->
- [x] verified that the documentation builds locally?
<!-- Please build the documentation (i.e., type make in the "doc" directory) and make sure it finishes. -->
- [ ] squashed/minimized your commits and written descriptive commit messages?
<!-- We value a clean history with useful commit messages. Ideally, you will take care of this
         before submitting a PR; otherwise you'll be asked to do so before merging. -->
- [x] added or updated existing tests to cover the changes?
- [ ] updated the documentation?
- [ ] added an example?

reneeotten commented:
thanks @newville for putting a PR together for this; I'll try to take closer look in the next few days.

newville commented:
@reneeotten Thanks for looking at this!  Is there any preference to try to get this into 0.9.15 or wait for v 1.0?

reneeotten commented:
> @reneeotten Thanks for looking at this! Is there any preference to try to get this into 0.9.15 or wait for v 1.0?

I have no preference; if it's done before the end of the week it can go in 0.9.15, otherwise not. It doesn't look like something that is likely to break downstream project, so I don't think it's that big of a deal. Let's try to wrap this up this week take it from there.

reneeotten commented on lmfit/minimizer.py:
```diff
@@ -875,9 +917,14 @@ def scalar_minimize(self, method='Nelder-Mead', params=None, **kws):
+        if 'maxiter' in kws:
+            warnings.warn(MAXEVAL_Warning % ('maxiter', thisfuncname()),
+                          RuntimeWarning)
+            kws.pop('maxiter')
```
this ```maxiter``` check with a method through ```scalar_minimize``` does not seem to work when using the Model interface

newville replied:
Will investigate... I think that ideally we could pop away `maxiter` for all the solvers.  But I'm not confident that we will find all of these....

reneeotten commented on lmfit/minimizer.py:
```diff
@@ -1457,19 +1508,20 @@ def least_squares(self, params=None, **kws):
         # here for consistency, and assuming it will be fixed.
-        if not result.aborted:
-            result.residual = self.__residual(ret.x, False)
+        # if not result.aborted:
```
unneeded leftover?

newville replied:
Yeah, good point.  I'll tidy that up.

reneeotten replied:
can you also comment on the earlier review comments I left; it would be helpful to know your idea behind it when I look more closely at the code. Thanks!

newville commented on lmfit/minimizer.py:
```diff
@@ -76,6 +77,17 @@
+        return inspect.stack()[1].function
+    except AttributeError:
+        return inspect.stack()[1][3]
+
```
yeah, that was the idea. it might need fleshing out.

newville commented on lmfit/minimizer.py:
```diff
@@ -1571,12 +1628,22 @@ def leastsq(self, params=None, **kws):
             _best, _cov, infodict, errmsg, ier = lsout
+        else:
+            _best = [self.result.params[p].value for p in self.result.var_names]
+            _best = np.array(_best)
```
OK. not sure why it's like that. Perhaps > 80 characters. Still, I agree with you and will fix.

newville commented on tests/test_itercb.py:
```diff
@@ -47,3 +47,20 @@ def test_itercb(method, calc_covar):
+    assert out.nfev < 20
+    assert out.aborted
+    assert not out.errorbars
+    assert not out.success
```
Yes, that's all probably true, and probably reflects the process of getting it to work at all rather than what should be the result. Will fix.

newville commented on lmfit/minimizer.py:
```diff
@@ -1457,19 +1508,20 @@ def least_squares(self, params=None, **kws):
-        result._calculate_statistics()
+            result.residual = self.__residual(ret.x, False)
 
+        result._calculate_statistics()
```
Basically, the problem was that many aborted fits end up without a valid `result.residual` array. That makes `_calculate_statistics` complain.  It seems like "too many func evals" really ought to return the residual so far and try to calculate the fit stats.  Obviously, the results won't be the best solution but it still seems useful to have.

newville commented on lmfit/minimizer.py:
```diff
@@ -1571,12 +1628,22 @@ def leastsq(self, params=None, **kws):
-            return result
+            result._calculate_statistics()
+        except AbortFitException:
+            pass
```
same as above... no valid `residual`.   I'll see if I can clean this up.

newville commented:
@reneeotten (sorry for the holiday interruption -- and yes, I am cooking and will be cooking++ all day tomorrow).  I *think* I've addressed your reviews.    For sure, the use of `max_nfev` in the global solvers `ampgo`, `shgo`, `brute`, `basinhopping`, and `de` might need more work.  But I think `scalar_minimize`, `leastsq`, and `least_squares` now work as expected and warn if the user provides  `maxfev` or `maxiter`.


In the process, I also noticed a separate issue. The tested "fit sine wave" example in `test_nose.py` and as in `doc_parameters_basic.py` was not calculating uncertainties with `numdifftools` when using `method='Nelder`'.  This seems to be from a poor default choice of the `step` with `numdifftools.Hessian'.  Using `Hessian(self.penalty, step=1.e-3)` fixed that (and gives pretty decent results!).   That meant the testing neede to be updated.

The commit history is sort of mess (using Travis to test works!).  This could be squashed or altered anyway you suggest, and a re-review would be most appreciated!

reneeotten commented:
> @reneeotten (sorry for the holiday interruption -- and yes, I am cooking and will be cooking++ all day tomorrow). I _think_ I've addressed your reviews.
>
> The commit history is sort of mess (using Travis to test works!). This could be squashed or altered anyway you suggest, and a re-review would be most appreciated!

Thanks @newville - I will certainly take a look at this sometime after the turkey ;) Happy Thanksgiving!

reneeotten commented:
@newville okay, this might not be what you want to hear... but, I am not sure if I really like not allowing an end-user to use the underlying ```scipy``` keywords (i.e., dropping ```maxfev```, ```maxiter``` and perhaps ```maxfun``` or others that are used). I think the user should be able to access this from within ```lmfit``` if he/she wants to. (Right now, the ```kws.pop``` doesn't seem to work in some cases where that keyword is actually specified in the ```options``` dictionary. At least that appears to be happening for a few cases that I looked at... I haven't investigated that further, since I am not sure we should go that route anyway).

How about we keep the ```max_nfev``` , but only set ```self.max_nfev``` (initialized right now as ```None```) to a value when the user specifies ```max_nfev``` in the ```Minimizer```, ```Model.fit()``` or when directly calling ```minimizer.<algorithm>```. In ```_residual``` we should  be able to do then:
```
if self.max_nfev and self.result.nfev > self.max_nfev:
    <abort code that you introduced there>
```

To me that would us to introduce a "general" way of specifying the "maximum number of function evaluations" that we can add to the documentation. At the same time, we keep all the flexibility and underlying ```scipy``` code accessible for people who want to us it. Let me know what you think or if there is any consideration I missed here... Again, since what I'm suggesting will be more work - I am happy to give that a try provided that you agree with the idea.

newville commented:
@reneeotten  Thanks. I think the concern in #332 and here is that the scipy API is inconsistent (possibly even "horribly" so) in the way to say "maximum number of evaluations of the objective functions" (and leaving aside +/- a couple).  There are `maxfev`, `max_nfev`, `maxfun`, and  `maxiter`.  This inconsistency means that the user needs to know the correct keyword argument for each method.    'maxiter` is particularly confusing because "number of function evaluations per iteration" is not defined clearly for any method. There is no uniform meaning of "iteration".  

The `scipy.optimize` API is inconsistent nd they are not going to change it for the better.  They may even continue to make it worse, as they most certainly did with `least_squares`, and to some extent with `minimize`.   Note that `max_nfev` was introduced into scipy.optimize by `least_squares` -- they already had at least 3 ways to say "max function evaluations" and they decided, reviewed, and approved adding a 4th way to say this.   Yes, they did.

This PR is a small attempt to add some consistency so that the API is not so insane as to have 4 ways of spelling the same concept.   I'm not sure that the user should be forbidden from setting both `max_nfev` and `maxfev`, etc, but it could certainly be confusing.     So I don't really see any point in supporting both.  I think we should either say there is a way to specify "max function evaluations"  in lmfit or we should continue saying "you have to drop down to the inconsistent, error-prone  API of the individual solvers.

I don't have a strong preference.  This was an effort to address #332 and to try to smooth over the pain of `scipy.optimize`.   But it is lower priority than releasing 0.9.15 soon (and probably with an rc1 that gets promoted to final in a matter of days) so we can branch it away and drop support for Python 2 as soon as possible.    Seriously, we are running out of time.    Merge or not?

reneeotten commented:
@newville, okay I see your point and can be convinced to go with the “max_nfev” approach as you have started in this PR. 

But, then it should work consistently with all solvers and with both the Minimizer and Model classes (also with respect to dropping the other scipy keywords). I am currently (after a few hours of working with it and attempting to write more tests) not convinced that this is the case as I already said above. 

So if you want to tag 0.9.15rc and release a version within a few days, then no merge. The ticket has been open for 3 years, so I think it’s worthwhile to spend a bit more time to work out the kinks. I’ll look at it again during the weekend.

So feel free to tag a 0.9.15rc, I can update some final documentation and add a “what’s new” on Sat/Sun and then we should be good to go. We can then duplicate that as a “python2” branch and start cleaning up for PY3-only on “master”. Hopefully by the time we’re ready for a 1.0 release, this PR is finalized as well.

newville commented:
@reneeotten OK, I will tag and push 0.9.15rc1, by Monday evening (US time), without merging this.

> But, then it should work consistently with all solvers and with both the Minimizer and Model classes (also with respect to dropping the other scipy keywords). I am currently (after a few hours of working with it and attempting to write more tests) not convinced that this is the case as I already said above.

I would caution against expecting completeness and perfection here.  This PR does attempt to rationalize the API.  `max_nfev` would be the documented way to express this concept.   As far as I know, it does work with the simpler, local solvers (leastsq, least_squares, minimize), and with Model.  I would not ever guarantee that it is perfect.  There may be issues to solve with global solvers, especially those that employ other local solvers or multi-processing.  Those issues currently exist too.

Merging an imperfect solution would give a starting point and a framework for solving the corner cases and adding consistency.   Not merging leaves no solution at all.

This PR does not attempt to rationalize any other scipy keywords.  This PR was actually some work and expecting it to rationalize other keywords would add more work.  Not accepting a PR because it is not extended far enough to other issues a sure-fire way to kill enthusiasm.  I would caution against doing that.  Invite solutions and build them up, slowly if needed.   

Like you say, this issue is long-standing.  It will not be included in 0.9.15.  I do not intend to work on this PR for a release of 1.0.  I have no expectations for how development will proceed after 1.0.  It is entirely possible that #332 will remain unresolved.

reneeotten commented:
@newville as I said earlier, I do appreciate the work you put into this PR and am not trying to diminish any of that or to kill enthusiasm. You asked my opinion and I gave it - that's all, I didn't mean to offend you in any way. 

I also promised to look at it more closely in the next few days (or perhaps week) to see if it's feasible to work out some of the issues I brought up. We can re-evaluate afterwards and see if it can be included in 1.0. I just think it's too early to merge the current PR into a released version.

newville commented:
@reneeotten  Oh,  I am not offended ;).   I agree this PR is sort of low priority and is an "attempt". It is definitely rushed.  So not merging this really is OK with me.   

But also: I expect that someone else would have to take over the work here to get something acceptable.   If or when that does happen, I would suggest being a bit more open to accepting PRs as incremental progress.

I'll leave this open for now, but have no intention of merging it.

patquem commented:
Hi everyone,
You worked a while on `max_fev` (thanks), and some things related to that argument are still unclear to me.
I  created the example below which illustrates some strange and confusing behaviour for me.
Here are some comments/questions that this example inspire me:
1 -  using `max_nfev` works fine in case of `leastsq` when `max_nfev` is reached
2 - using `leasts_squares` : a/ `nfev` is abnormally low compared to those obtained with `leastsq` b/ when `max_nfev` seems to be reached, `nfev` radically changes of value (to be more conform with the "real" and expected `nfev` I think)
3 - `least_square` (without an 's' - that is not referenced) returns the same results than 'least_squares'
4 - A non-existing fitting method ('xxxxx') returns nevertheless a results (??? from which algorithm ???)
Thank you for your help,
Patrick

lmfit version : 1.0.3

```
import numpy as np
import matplotlib.pyplot as plt
from lmfit import Model
from lmfit.models import LorentzianModel

x = np.arange(200)
y = np.sin(0.2 * x).clip(0)


def eval(method, max_nfev=None):
    label = f"{method} - max_nfev={max_nfev}"
    print(label)
    plt.figure()
    plt.title(label)
    plt.plot(x, y)
    for i in range(4):
        model = LorentzianModel(prefix=f'm{i}_')
        model.set_param_hint("amplitude", min=0)
        model.set_param_hint("sigma", max=10)
        model.set_param_hint("center", value=40 * (i + 1))
        if i == 0:
            models = model
        else:
            models += model
        out = models.fit(y, x=x, method=method, max_nfev=max_nfev)
        print(i, out.nfev)
        plt.plot(x, out.best_fit, label=f'fit-{i + 1}')
    plt.legend()
    print()


eval(method='leastsq', max_nfev=None)
eval(method='leastsq', max_nfev=150)
eval(method='least_squares', max_nfev=None)
eval(method='least_squares', max_nfev=150)

eval(method='least_square', max_nfev=None)
eval(method='xxxxx', max_nfev=None)

plt.show()
```
```
leastsq - max_nfev=None
0 61
1 73
2 133
3 212

leastsq - max_nfev=150
0 61
1 73
2 133
3 150

least_squares - max_nfev=None
0 11
1 11
2 17
3 14

least_squares - max_nfev=150
0 11
1 11
2 151
3 151

least_square - max_nfev=None
0 11
1 11
2 17
3 14

xxxxx - max_nfev=None
0 170
1 1513
2 2787
3 5366
```

newville commented:
@patquem Commenting on PRs that were merged long ago makes it really challenging for us to work with.  Like, do you expect us to change this PR nearly 3 years after it was merged?  That's just not going to happen. 

If you want to raise an issue about the current state of the code, well, raise an issue.  I agree that the content of your message needs to be looked into.

patquem commented:
Hi @newville ,
The purpose of my comment was certainly not to change the PR.
As it related to what was discussed here, I thought appropriate to write it after that.
No problem, I will create a new issue :)
